### PR TITLE
[Peer Review] Hadoop/Hive

### PR DESCRIPTION
**Pull request by aayushi93**
Ticket URL -Hive Project -  https://www.notion.so/Hive-Project-3f18168fbc60494e82d271616c388b97
Ticket URL - Hadoop/Hive README - https://www.notion.so/README-md-c2c71eea64634478b75117b2ee9bd35e

### CODE REVIEW

File: hadoop/README.md [hunk 1]
```diff
@@ -1,3 +1,85 @@
...
+- __DataNode__: Also known as worker node. These are workhorses of the filesystem. They store and retrieve blocks when the are told to do so by the namenode. The report back to the namenode periodically, with list of blocks they are storing.
+	
+### MapReduce
+MapReduce is a programming model that allows developers to process large amount of data. MapReduce job processes data in 3 phases: The Mapping phase, The Shuffle phase and the reduce phase. 
+- __The Map Phase__:  This phase consists of mappers that are user defined functions. It takes a part of data and applies business logic to it. The input data is converted into key-value pairs based on the mappers. This part is performed by data nodes. 
+- __The Shuffle Phase__:  In this phase, the data is passed to the Reducers from the mappers. 
+- __The Reduce Phase__:  In this phase, data is received from the mappers. Aggregation, filtering, sorting or combination of operations are applied on the data based on the requirements. The output of this reducer that is located at the namenode will be the final result.
```
Comment (vassairm) [inline hadoop/README.md:40]: I would double-check this. There wont always be a reduce phase to begin with, and I'm not sure if containers are ever run on the Master.

File: hadoop/README.md [hunk 2]
```diff
@@ -1,3 +1,85 @@
...
+	
+### YARN
+Apache YARN (Yet Another Resource Negotiator) is Hadoops cluster resource management system. It was introduced to improve MapReduce implementation and support other distributed computing paradigms as well. YARN provides its core services via two long-running daemons: The Resource Manager and The Node Managers. 
+- __The Resource Manager__:  Each cluster has one resource manager to manage the use of resources across the cluster and assign resources to different jobs. Resource manager runs on master node and keeps track of workers location and their hardware usage.
+- __Node Manager__:  It runs on all the worker nodes in the cluster to launch and monitor containers. It gathers the hardware usage information and send it to resource manager.
```
Comment (vassairm) [inline hadoop/README.md:47]: Mention heartbeats, and maybe Application Masters as well

File: hadoop/README.md [hunk 3]
```diff
@@ -1,3 +1,85 @@
...
+- __Zeppelin Notebook__:  Zeppelin Notebooks provide an easy way to execute code in a wb notebook. Zeppelin can be used to execute scala, SQL, ssh commands and even schedule a job with cron to run at regular intervals. 
+Using Beeline CLI, it is difficult to manage, execute and present large number of queries. This is one of the advantages of using Zeppelin notebooks. It helps us manage and present queries in a systematic way as the queries are organized into pages and paragraphs.  
+- __Zeppelin Configs__:  Zeppeling runs on master node to establish Hive connection. A customized Hive interpreter is configured based o default JDBC interpreter. 
+- __Zeppelin Usage__:   Zeppelin notebook, in our case, is organized into different paragraphs, each paragraph performing different HiveQL queries. There are few where ssh commands are run after putting %sh at the top of the paragraph. Each paragraph on the page is executed individually and sequentially.
```
Comment (vassairm) [inline hadoop/README.md:64]: Zeppelin is separate from hive

File: hadoop/README.md [hunk 4]
```diff
@@ -1,3 +1,85 @@
...
+- __The Resource Manager__:  Each cluster has one resource manager to manage the use of resources across the cluster and assign resources to different jobs. Resource manager runs on master node and keeps track of workers location and their hardware usage.
+- __Node Manager__:  It runs on all the worker nodes in the cluster to launch and monitor containers. It gathers the hardware usage information and send it to resource manager. 
+
+### HIVE
+Apache Hive is a framework for data warehousing on top of Hadoop. It facilitates reading, writing, and managing large datasets residing in the distributed storage using SQL. Hive provides easy access to data via SQL, enabling data warehousing tasks such as Extract-Transform-Load (ETL), reporting and data analysis. HIVE has 2 components: Hcatalog and WebHCat.
+- __Hcatalog__:  Hcatalog is a table and storage management layer for Hadoop that enables users with different data processing tools  including Pig and MapReduce to more easily read and write data on the grid.
```
Comment (vassairm) [inline hadoop/README.md]: Since you mentioned HCatalog, you should mention the MetaStore.

File: hadoop/README.md [hunk 5]
```diff
@@ -1,3 +1,85 @@
...
+2. Perform Query optimizations.
+3. Copy data between tables.  
+4. Using Serializer/Deserialize (SerDe) to read data from a tabke and write it back to the HDFS in any custom format. SerDe interface allows user to instruct Hive on how a record should be processed.
+5. Partitioning tables into related parts based on the values of columns.
+6. Performance comparison between SparkQL and HiveQL.
+	 
+### Zeppelin Notebook
+Below is the screenshot of Zeppelin Notebook explaining tasks likes Reading data from Google BigQuery, copying data between tables, loading data into HDFS, creating partitions and comparing for efficiency and various other performance comparisions.
+
+![Zeppelin Notebook Screenshot](./Assets/Zeppelin_Hive.png)
```
Comment (vassairm) [inline hadoop/README.md]: I don't see where you make the first Hive table, wdi_gs.
Some of your markdown paragraphs aren't rendered.
You can safely skip the select * on the debug table.
You can skip Spark SQL show tables

File: hadoop/README.md [hunk 6]
```diff
@@ -1,3 +1,92 @@
...
+
+### Cluster Architecture Diagram
+
+![Hadoop Cluster Architecture](./Assets/Hadoop_Architecture.png)
+
+### Hadoop Framework
+Apache Hadoop is an open source framework written in Java. It allows distributed processing of massive amounts of data across multiple clusters of computers. The Hadoop Framework application provides a software framework for distributed storage and processing using MapReduce programming model. The storage part of Hadoop is known as Hadoop Distributed File System (HDFS). In 2010, YARN was introduced that manages computing resources in clusters and utilizing them for scheduling users application. 
+
+### HDFS
+HDFS is a filesystem designed for storing very large files with streaming data access patterns that run on clustersw of commodity hardware. It is built around the idea that most efficient data processing pattern is write-once and read-many times pattern. HDFS cluster has two types of nodes operating in a master-slave pattern, as follows:
```
Comment (yangdav9) [inline hadoop/README.md]: runs on clusters* of commodity hardware

File: hadoop/README.md [hunk 7]
```diff
@@ -1,3 +1,92 @@
...
+- __NameNode__: Also known as Master node. The namenode manages the filesystem namespace. It maintains the filesystem tree and the metadata for all the files and directories in the tree. The name node also knows the datanodes on which all the blocks for a given file are located.
+- __DataNode__: Also known as worker node. These are workhorses of the filesystem. They store and retrieve blocks when the are told to do so by the namenode. The report back to the namenode periodically, with list of blocks they are storing.
```
Comment (yangdav9) [inline hadoop/README.md:34]: Can use term 'Heartbeat' for datanode report to namenode

**Pull request merged**